src/nwb_benchmarks/benchmarks/streaming_base.py

```diff
@@ -0,0 +1,101 @@
+"""Base template classes for basic benchmarks for NWB streaming read."""
+import fsspec
+import h5py
+import pynwb
+import remfile
+from fsspec.asyn import reset_lock
+
+
+class FileReadStreamingBase:
+    """Base class for basic benchmarks for streaming an NWB file from the DANDI archive."""
+    def setup(self):
+        # Random IBL raw data file; not that many groups
+        self.s3_url = "https://dandiarchive.s3.amazonaws.com/blobs/8c5/65f/8c565f28-e5fc-43fe-8fb7-318ad2081319"
+
+    def fsspec_no_cache(self):
+        reset_lock()
+        fsspec.get_filesystem_class("https").clear_instance_cache()
+        filesystem = fsspec.filesystem("https")
+
+        with filesystem.open(path=self.s3_url, mode="rb") as byte_stream:
+            with h5py.File(name=byte_stream) as file:
+                with pynwb.NWBHDF5IO(file=file, load_namespaces=True) as io:
+                    nwbfile = io.read()
+
+        # Must be done at this level since teardown occurs outside of repetitions
+        # reset_lock()
+        # fsspec.get_filesystem_class("https").clear_instance_cache()
+
+    def ros3(self):
+        ros3_form = self.s3_url.replace("https://dandiarchive.s3.amazonaws.com", "s3://dandiarchive")
+        with pynwb.NWBHDF5IO(path=ros3_form, mode="r", load_namespaces=True, driver="ros3") as io:
+            nwbfile = io.read()
+
+    def remfile(self):
+        byte_stream = remfile.File(url=self.s3_url)
+        with h5py.File(name=byte_stream) as file:
+            with pynwb.NWBHDF5IO(file=file, load_namespaces=True) as io:
+                nwbfile = io.read()
+
+
+class ElectricalSeriesStreamingROS3Base:
+    """
+    "Base class for basic benchmark for streaming raw ecephys data.
+
+    Needs separate setup per class to only time slicing operation.
+    """
+    def setup(self):
+        self.s3_url = "s3://dandiarchive/blobs/8c5/65f/8c565f28-e5fc-43fe-8fb7-318ad2081319"
+        self.acquisition_path = "ElectricalSeriesAp"
+        self.slice_range = (slice(0, 30_000), slice(0, 384))  # ~23 MB
+        self.io = pynwb.NWBHDF5IO(path=self.s3_url, mode="r", load_namespaces=True, driver="ros3")
+        self.nwbfile = self.io.read()
```

Comment:
Likewise, I'd propose separate this even further into

```python
ElectricalSeriesStreamingBase:
    ...
    acquisition_path: str  # name of object in nwbfile.acquisition
    slice_range: Tuple[slice, ...]

    def setup();
        assert self.acquisition_path is not None, "You must specify the name of the ElectricalSeries in the acquisition module."
        assert self.slice_range is not None, "You must specify the slice of the electrical series data."

    def time_slice_request():
        self.nwbfile.acquisition[self.acquisition_path].data[self.slice_range]
```

so that child classes define different setups

```python
class ElectricalSeriesROS3Base(ElectricalSeriesStreamingBase):
    ...

    def setup():
        super().setup()
        assert self.s3_url is not None, "..."

        self.io = pynwb.NWBHDF5IO(path=self.s3_url, mode="r", load_namespaces=True, driver="ros3")
        self.nwbfile = self.io.read()
```

and actual instances in the test suite just specify the minimal args to make it explicit what is being tested

```python
class ElectricalSeriesStreamingROS3Case1(ElectricalSeriesROS3Base):
    s3_url = "some asset"
    self.acquisition_path = "ElectricalSeriesAp"
    self.slice_range = (slice(0, 30_000), slice(0, 384))  # ~23 MB
```

but obscure the other details of _how_ ROS3/fsspec/remfile setup the `nwbfile` object

Reply:
I implemented these changes. In addition I only defined the logic for the test cases in the base class but left it to the actual instance to call it to allow the same test case to be run with different metrics.